notebooks/chat.ipynb

```diff
@@ -29,892 +67,491 @@
     "# Multi-step chat with hidden blocks"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We are now going to set up a multistage chat, where we have the chat bot help the use achieve some goal.\n",
+    "The user will only have to specify the goal, and then we will create a chain-of-thought conversation with the bot which will:\n",
+    "\n",
+    "1. Ask the bot for a number of suggestions\n",
+    "2. List the pros and cons of each\n",
+    "3. Pick the best suggestion\n",
+    "4. Product a detailed action plan\n",
+    "\n",
+    "Our goal is to only show the final result to the user."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, let us define our generation function:"
+   ]
+  },
   {
    "cell_type": "code",
-   "execution_count": 8,
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "\n",
+    "import guidance\n",
+    "from guidance import system, user, assistant\n",
+    "\n",
+    "@guidance\n",
+    "def plan_for_goal(lm, goal: str):\n",
+    "    # This is a helper function which we will use below\n",
+    "    def parse_best(prosandcons, options):\n",
+    "        best = re.search(r'Best=(\\d+)', prosandcons)\n",
+    "        if not best:\n",
+    "            best =  re.search(r'Best.*?(\\d+)', 'Best= option is 3')\n",
+    "        if best:\n",
+    "            best = int(best.group(1))\n",
+    "        else:\n",
+    "            best = 0\n",
+    "        return options[best]\n",
+    "\n",
+    "    # Some general instruction to the model\n",
+    "    with system():\n",
+    "        lm += \"You are a helpful assistant.\"\n",
+    "\n",
+    "    # Simulate a simple request from the user\n",
+    "    # Note that we switch to using 'lm2' here, because these are intermediate steps\n",
```

Comment:
Looking at @marcotcr 's notebook @slundberg , perhaps I should use `with silent()` instead?